### PR TITLE
Consolidate load functions, add schema for casting, and add CSV options for reading

### PR DESCRIPTION
**Pull request by arienandalibi**
Takes care of #2394 

# Branch `consolidate_load_functions` - API Changes Summary

This document summarizes the API changes in the `consolidate_load_functions` branch compared to `master`.

## Overview

This branch consolidates multiple data loading functions into unified methods that accept any Arrow-compatible data source, file paths, or directories.

---

### 1. Unified Load Functions

**Old API (removed):**
```python
# These functions have been removed:
g.load_edges_from_pandas(df, ...)
g.load_edges_from_parquet(path, ...)
g.load_edge_props_from_pandas(df, ...)
g.load_edge_props_from_parquet(path, ...)
g.load_nodes_from_pandas(df, ...)
g.load_nodes_from_parquet(path, ...)
g.load_node_props_from_pandas(df, ...)
g.load_node_props_from_parquet(path, ...)

# PersistentGraph-specific (also removed):
g.load_edge_deletions_from_pandas(df, ...)
g.load_edge_deletions_from_parquet(path, ...)
```

**New API (unified):**
```python
# Now there are 4 unified functions:
g.load_edges(data, time, src, dst, ...)
g.load_nodes(data, time, id, ...)
g.load_edge_metadata(data, src, dst, ...)
g.load_node_metadata(data, id, ...)

# PersistentGraph-specific:
g.load_edge_deletions(data, time, src, dst, ...)
```

### 2. Multi-Source Data Input

The `data` parameter accepts any of the following:
- **Pandas DataFrames**
- **Polars DataFrames**
- **FireDucks DataFrames** (`fireducks.pandas`)
- **PyArrow Tables**
- **DuckDB query results** (`DuckDBPyRelation`)
- **File paths** (string or `Path` object):
  - Single CSV file
  - Single Parquet file
  - Directory containing CSV/Parquet files
  - Compressed files (`.csv.gz`, `.csv.bz2`)

**Examples:**
```python
from raphtory import Graph
import pandas as pd
import polars as pl

g = Graph()

# From Pandas DataFrame
df_pd = pd.DataFrame({"time": [1,2], "src": [1,2], "dst": [2,3], "value": [10,20]})
g.load_edges(data=df_pd, time="time", src="src", dst="dst", properties=["value"])

# From Polars DataFrame
df_pl = pl.DataFrame({"time": [1,2], "src": [1,2], "dst": [2,3], "value": [10,20]})
g.load_edges(data=df_pl, time="time", src="src", dst="dst", properties=["value"])

# From file path
g.load_edges(data="/path/to/data.parquet", time="time", src="src", dst="dst")
g.load_edges(data="/path/to/data.csv", time="time", src="src", dst="dst")

# From directory (loads all CSV/Parquet files)
g.load_edges(data="/path/to/directory/", time="time", src="src", dst="dst")

# From compressed CSV
g.load_nodes(data="/path/to/data.csv.gz", time="time", id="node_id")
```

### 3. Schema Parameter for Column Casting

New `schema` parameter allows explicit type casting of columns:

```python
from raphtory import Graph, PropType
import pyarrow as pa

g = Graph()

# Using PropType
g.load_nodes(
    data=df,
    time="time",
    id="id",
    properties=["value"],
    schema=[("value", PropType.i64())]  # Cast to int64
)

# Using PyArrow DataType
g.load_nodes(
    data=df,
    time="time",
    id="id",
    properties=["timestamp"],
    schema={"timestamp": pa.timestamp("ms", tz="UTC")}  # Cast to datetime
)

# Using dict format
g.load_nodes(
    data=df,
    time="time", 
    id="id",
    properties=["value"],
    schema={"value": PropType.i64()}
)

# Nested types (lists, maps)
g.load_nodes(
    data=table,
    time="time",
    id="id",
    properties=["values"],
    schema=[("values", PropType.list(PropType.i64()))]
)
```

### 4. CSV Options Parameter

New `csv_options` parameter for CSV reading configuration:

```python
g = Graph()

g.load_nodes(
    data="/path/to/semicolon_data.csv",
    time="time",
    id="id",
    csv_options={
        "delimiter": ";",        # Delimiter character
        # Other available options:
        "comment": "#",        # Comment character
        "escape": "\\",        # Escape character
        "quote": '"',          # Quote character
        "terminator": "\n",    # Line terminator
        "allow_truncated_rows": True,  # Allow rows with fewer columns
        "has_header": True,    # Whether CSV has header row
    }
)
```

### `Graph.load_edges()`
```python
def load_edges(
    self,
    data: Any,                          # DataFrame, file path, or directory
    time: str,                          # Timestamp column name
    src: str,                           # Source node column name
    dst: str,                           # Destination node column name
    properties: Optional[List[str]] = None,
    metadata: Optional[List[str]] = None,
    shared_metadata: Optional[PropInput] = None,
    layer: Optional[str] = None,
    layer_col: Optional[str] = None,
    schema: Optional[list[tuple[str, DataType | PropType | str]] | dict[str, DataType | PropType | str]] = None,
    csv_options: Optional[dict[str, str | bool]] = None
) -> None
```

### `Graph.load_nodes()`
```python
def load_nodes(
    self,
    data: Any,
    time: str,
    id: str,
    node_type: Optional[str] = None,
    node_type_col: Optional[str] = None,
    properties: Optional[List[str]] = None,
    metadata: Optional[List[str]] = None,
    shared_metadata: Optional[PropInput] = None,
    schema: Optional[list[tuple[str, DataType | PropType | str]] | dict[str, DataType | PropType | str]] = None,
    csv_options: Optional[dict[str, str | bool]] = None
) -> None
```

### `Graph.load_edge_metadata()`
```python
def load_edge_metadata(
    self,
    data: Any,
    src: str,
    dst: str,
    metadata: Optional[List[str]] = None,
    shared_metadata: Optional[PropInput] = None,
    layer: Optional[str] = None,
    layer_col: Optional[str] = None,
    schema: Optional[list[tuple[str, DataType | PropType | str]] | dict[str, DataType | PropType | str]] = None,
    csv_options: Optional[dict[str, str | bool]] = None
) -> None
```

### `Graph.load_node_metadata()`
```python
def load_node_metadata(
    self,
    data: Any,
    id: str,
    node_type: Optional[str] = None,
    node_type_col: Optional[str] = None,
    metadata: Optional[List[str]] = None,
    shared_metadata: Optional[PropInput] = None,
    schema: Optional[list[tuple[str, DataType | PropType | str]] | dict[str, DataType | PropType | str]] = None,
    csv_options: Optional[dict[str, str | bool]] = None
) -> None
```

---

### `Properties.get_dtype_of()`

Get the PropType of a property. Returns the PropType of the latest value for this property if it exists. If not, it returns the PropType for the static property matching this name.

```python
def get_dtype_of(key: str) -> Optional[PropType]
```

**Arguments:**
- `key` (str): the name of the property.

**Returns:**
- `PropType`: The type of the property, or `None` if not found.

**Example:**
```python
node = g.node("Alice")
dtype = node.properties.get_dtype_of("age")
print(dtype)  # PropType.i64()
```

---

## PersistentGraph-Specific

`PersistentGraph` also includes:
```python
def load_edge_deletions(
    self,
    data: Any,
    time: str,
    src: str,
    dst: str,
    layer: Optional[str] = None,
    layer_col: Optional[str] = None,
    schema: Optional[...] = None,
    csv_options: Optional[dict[str, str | bool]] = None
) -> None
```

### CODE REVIEW

File: python/tests/test_load_from_df.py [hunk 1]
```diff
@@ -39,6 +43,739 @@ def test_load_edges_from_polars_df(graph_type):
...
+
+    # no casting
+    g = Graph()
+    g.load_nodes(
+        data=table,
+        time="time",
+        id="id",
+        properties=["val_struct"],
+    )
+    d_type_no_cast = g.node(10).properties.get_dtype_of("val_struct")
```
Comment (miratepuffin) [inline python/tests/test_load_from_df.py:358]: This is a new function no?

File: python/tests/test_load_from_df.py [hunk 2]
```diff
@@ -39,6 +43,739 @@ def test_load_edges_from_polars_df(graph_type):
...
+            with pytest.raises(Exception, match="Null not supported for time column"):
+                g = Graph()
+                g.load_nodes(
+                    data=malformed_file,
+                    time="block_timestamp",
+                    id="inputs_address",
+                    properties=["outputs_address"],
+                )
+
+        if "null_timestamp.parquet" in malformed_file.name:
```
Comment (miratepuffin) [inline python/tests/test_load_from_df.py:710]: We can add after merge, but we probably need to add similar tests to these but for load_edges delete_edges and edge/node metadata maybe?

File: raphtory-api/src/python/prop.rs [hunk 3]
```diff
@@ -149,8 +155,8 @@ impl PyProp {
         Ok(PyProp(Prop::Map(Arc::new(map))))
     }
 
-    pub fn dtype(&self) -> String {
-        format!("{:?}", self.0.dtype())
+    pub fn dtype(&self) -> PropType {
```
Comment (miratepuffin) [inline raphtory-api/src/python/prop.rs:158]: I think the new one makes sense, but probably best we all give it a thumbs up

File: raphtory-api/src/python/prop.rs [hunk 4]
```diff
@@ -228,3 +234,143 @@ impl<'source> FromPyObject<'source> for Prop {
...
+        PyPropType(self).into_pyobject(py)
+    }
+}
+
+impl<'source> FromPyObject<'source> for PropType {
+    fn extract_bound(ob: &Bound<'source, PyAny>) -> PyResult<Self> {
+        if let Ok(prop_type) = ob.downcast::<PyPropType>() {
+            Ok(prop_type.get().0.clone())
+        } else if let Ok(prop_type_str) = ob.extract::<PyBackedStr>() {
+            match prop_type_str.deref().to_ascii_lowercase().as_str() {
```
Comment (miratepuffin) [inline raphtory-api/src/python/prop.rs:350]: Do we need decimal here?

File: raphtory/src/python/graph/io/arrow_loaders.rs [hunk 5]
```diff
@@ -254,33 +298,531 @@ pub(crate) fn process_arrow_c_stream_df<'a>(
...
+        Ok(x) => {
+            if let Ok(x) = x.extract() {
+                kdam::set_notebook(x);
+            }
+        }
+        Err(e) => {
+            error!("Error checking if running in a jupyter notebook: {}", e);
+        }
+    };
+}
```
Comment (miratepuffin) [inline raphtory/src/python/graph/io/arrow_loaders.rs:828]: Are some of these helper functions also useful in rust?

**Review by miratepuffin — COMMENTED** [review]
Couple of minor comments that are probably not for this PR:
Got extensive testing for node loaders, probably need something similar for edges, metadata and deletions.
Quite a few IO helpers in python that possibly could be useful in rust loading?
Need to fix one more doc test before we merge

File: raphtory/src/python/graph/graph.rs [hunk 6]
```diff
@@ -640,158 +642,129 @@ impl PyGraph {
...
-    ///     src (str): The column name for the source node ids.
-    ///     dst (str): The column name for the destination node ids.
+    ///     src (str): The column name for the source node IDs.
+    ///     dst (str): The column name for the destination node IDs.
     ///     properties (List[str], optional): List of edge property column names. Defaults to None.
     ///     metadata (List[str], optional): List of edge metadata column names. Defaults to None.
     ///     shared_metadata (PropInput, optional): A dictionary of metadata properties that will be added to every edge. Defaults to None.
     ///     layer (str, optional): A value to use as the layer for all edges. Cannot be used in combination with layer_col. Defaults to None.
     ///     layer_col (str, optional): The edge layer column name in a dataframe. Cannot be used in combination with layer. Defaults to None.
+    ///     schema (list[tuple[str, DataType | PropType | str]] | dict[str, DataType | PropType | str], optional): A list of (column_name, column_type) tuples or dict of {"column_name": column_type} to cast columns to. Defaults to None.
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/graph.rs:756]: We should probably add this as a `SchemaInput` alias in typing?

File: raphtory/src/python/graph/properties/props.rs [hunk 7]
```diff
@@ -98,6 +101,18 @@ impl PyProperties {
         self.props.get(key)
     }
 
+    /// Get the PropType of a property. Specifically, returns the PropType of the latest value for this property if it exists.
+    /// If not, it returns the PropType for the static property matching this name.
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/properties/props.rs]: this doesn't make sense? We don't have the fallback behaviour anymore

File: raphtory-api/src/python/prop.rs (hunk 3 above)
Comment (ljeub-pometry) [reply raphtory-api/src/python/prop.rs:158]: 👍
